### PR TITLE
fix: use object dtype for ARRAY columns in `to_pandas()` with pandas 1.x

### PR DESCRIPTION
**Pull request by tswast**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes internal issue 321013333
🦕

### CODE REVIEW

File: tests/system/small/ml/test_preprocessing.py [hunk 1]
```diff
@@ -580,8 +597,12 @@ def test_one_hot_encoder_different_data(penguins_df_default_index, new_penguins_
                 [{"index": 2, "value": 1.0}],
             ],
         },
-        dtype=pd.ArrowDtype(
-            pa.list_(pa.struct([("index", pa.int64()), ("value", pa.float64())]))
+        dtype=(
+            pd.ArrowDtype(
```
Comment (shobsi) [inline tests/system/small/ml/test_preprocessing.py:601]: seems to be a repetitive pattern, might be worth abstracting away in a method `get_compatible_dtype` inside `tests.system.utils`, or even `PANDAS_VERSIONS`

File: bigframes/session/_io/pandas.py [hunk 2]
```diff
@@ -74,6 +76,18 @@ def arrow_to_pandas(
                 else mask.to_numpy(zero_copy_only=False),
             )
             series = pandas.Series(pd_array, dtype=dtype)
+        elif (
+            isinstance(dtype, pandas.ArrowDtype)
```
[the remark this answers is not in the document]
Comment (tswast) [reply bigframes/session/_io/pandas.py:80]: I'd rather avoid too deep of nesting if we can. In this case, the more specific check is appearing before the more general check, as it should.

[the remark this answers is not in the document]
Comment (tswast) [reply bigframes/session/_io/pandas.py:80]: go/tott/651

[the remark this answers is not in the document]
Comment (tswast) [reply bigframes/session/_io/pandas.py:80]: That said, I might want to take that Testing on the Toilet's advice and create a helper function here.

File: tests/system/small/ml/test_preprocessing.py (hunk 1 above)
Comment (tswast) [reply tests/system/small/ml/test_preprocessing.py]: In this case the dtype is exactly the same. No need for a function. I just created a constant.

**Review by shobsi — APPROVED** [review]

File: bigframes/session/_io/pandas.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shobsi) [reply bigframes/session/_io/pandas.py]: Thanks!

**Comment by tswast** [comment]
Doctest failed with

```
2631             >>> @bpd.remote_function([str], bool, reuse=False)
UNEXPECTED EXCEPTION: InternalServerError('Could not create Cloud Run service bigframes-597cc02ef5ce0525e4f51697b5a83b6c-32er4jqu. ')
Traceback (most recent call last):
  File "/tmpfs/src/github/python-bigquery-dataframes/.nox/doctest/lib/python3.11/site-packages/google/api_core/grpc_helpers.py", line 79, in error_remapped_callable
    return callable_(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmpfs/src/github/python-bigquery-dataframes/.nox/doctest/lib/python3.11/site-packages/grpc/_channel.py", line 1160, in __call__
    return _end_unary_response_blocking(state, call, False, None)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmpfs/src/github/python-bigquery-dataframes/.nox/doctest/lib/python3.11/site-packages/grpc/_channel.py", line 1003, in _end_unary_response_blocking
    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.INTERNAL
	details = "Could not create Cloud Run service bigframes-597cc02ef5ce0525e4f51697b5a83b6c-32er4jqu. "
	debug_error_string = "UNKNOWN:Error received from peer ipv4:74.125.126.95:443 {grpc_message:"Could not create Cloud Run service bigframes-597cc02ef5ce0525e4f51697b5a83b6c-32er4jqu. ", grpc_status:13, created_time:"2024-01-22T17:51:55.118019527+00:00"}"
>
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/doctest.py", line 1351, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest series.Series.mask[7]>", line 1, in <module>
  File "/tmpfs/src/github/python-bigquery-dataframes/bigframes/functions/remote_function.py", line 851, in wrapper
    rf_name, cf_name = remote_function_client.provision_bq_remote_function(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmpfs/src/github/python-bigquery-dataframes/bigframes/functions/remote_function.py", line 472, in provision_bq_remote_function
    cf_endpoint = self.create_cloud_function(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmpfs/src/github/python-bigquery-dataframes/bigframes/functions/remote_function.py", line 415, in create_cloud_function
    operation = self._cloud_functions_client.create_function(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmpfs/src/github/python-bigquery-dataframes/.nox/doctest/lib/python3.11/site-packages/google/cloud/functions_v2/services/function_service/client.py", line 1007, in create_function
    response = rpc(
               ^^^^
  File "/tmpfs/src/github/python-bigquery-dataframes/.nox/doctest/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py", line 131, in __call__
    return wrapped_func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmpfs/src/github/python-bigquery-dataframes/.nox/doctest/lib/python3.11/site-packages/google/api_core/grpc_helpers.py", line 81, in error_remapped_callable
    raise exceptions.from_grpc_error(exc) from exc
google.api_core.exceptions.InternalServerError: 500 Could not create Cloud Run service bigframes-597cc02ef5ce0525e4f51697b5a83b6c-32er4jqu. 
/[tmpfs/src/github/python-bigquery-dataframes/third_party/bigframes_vendored/pandas/core/series.py:2631](https://cs.corp.google.com/piper///depot/google3/tmpfs/src/github/python-bigquery-dataframes/third_party/bigframes_vendored/pandas/core/series.py?l=2631): UnexpectedException
...
FAILED third_party/bigframes_vendored/pandas/core/frame.py::frame.DataFrame.map
FAILED third_party/bigframes_vendored/pandas/core/series.py::series.Series.apply
FAILED third_party/bigframes_vendored/pandas/core/series.py::series.Series.mask
```

As these are unrelated and `remote_function` is notoriously flaky, I'll override this check for now.

**Pull request merged**